text/000-rustdoc-jump-to-definition.md

```diff
@@ -0,0 +1,187 @@
+
+rust-analyzer in the browser: Another possibility would be to run it with wasm. However, it brings a few issues: it requires JS to work (obviously, but it's not a big problem), it would be quite heavy to run and it would complexify rustdoc build quite a lot.
+
+External library (like `cpan` or `tree-sitter`): They would work to generate links to another source code location. However, they bring their downsides as well: they would need to be kept up-to-date with rust syntax evolution (if any) and making links back to documentation would be very tricky because some information is only internal to rustdoc (like how to differentiate between different `impl blocks` or different blanket trait implementations like `impl Trait<T: Clone> for T` and `impl Trait<T: Copy> for T`).
+
+# Prior art
+[prior-art]: #prior-art
+
+As mentionned above, it is actually quite common for documentation tools to have a source viewer. Implementations vary a bit between them:
+
+| tool/language | source code viewer | jump to definition | go back to documentation |
+| ------------- | ------------------ | ------------------ | ------------------------ |
+| [doxygen](https://eigen.tuxfamily.org/dox/Matrix_8h_source.html) | yes | no | yes |
+| [Elixir](https://elixir.bootlin.com) (used for linux kernel) | yes | yes | no |
+| [Go](https://cs.opensource.google/go/go/+/go1.16.5:src/database/sql/sql.go;l=44) | yes | yes | no |
+| [hexdocs](https://preview.hex.pm/preview/k6/0.0.1/show/lib/k6/template/web_socket.ex) | yes | no | no |
+| python (sphynx/pydoc) | no (links to github) | no | no |
+| [ruby](https://ruby-doc.org/stdlib-3.1.2/libdoc/delegate/rdoc/Delegator.html#method-i-__raise__) | yes (directly on the page) | yes/no (it is sometimes in examples) | no |
+| [perl](https://metacpan.org/dist/DBD-SQLite/source/lib/DBD/SQLite.pm) | yes | yes | yes |
+| [haskell](https://devdocs.io/haskell~9/libraries/binary-0.8.9.0/data-binary-builder#v:append) | yes | yes | no |
+| [opal](https://opalrb.com/docs/api/v1.4.1/stdlib/Buffer) | yes (directly on the page) | no | yes |
+| [cargo-src](https://github.com/rust-dev-tools/cargo-src) | yes | yes | yes |
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+Is it the best way to make the links to go to an item documentation like this? Maybe something adding a small icon to click right besides it should be better?
+
+# Potential extensions
+
+A potential extension would be "find references". It would allow users to check where the given item is being used. We already have all the information needed for this feature, so it's mostly how to use it and show it to the end users. However, if we decide to move forward with it, it will be part of another RFC.
```

Comment:
this is exactly what i'm worried about wrt scope expansion, fwiw.

Rustdoc is not a source code browser tool. It has source in its capacity to _aid_ its purpose as a doc viewer. It is still insufficiently motivated (in this RFC) that better source code browsing is important for documentation viewing, because source is already very much a secondary purpose of rustdoc that it uses as an aid.

Going from there to reverse-search seems like an even bigger leap, and it only makes sense _if_ we have presupposed that Rustdoc is equally a tool for browsing source. And if we are deciding that, we need to decide that _very clearly_ in this RFC, and not have it just happen by accident.

The worry I have is that we are turning rustdoc from a tool designed for the _users_ of a crate to a tool designed for the _maintainers_ of a crate (and still the users, but now it's both). And that's a pretty big scope expansion.

Reply:
Just like prior-art, I added it because it was mentioned as a possible extension. Accepting this RFC doesn't necessarily mean that the feature will evolve any further. But if you prefer that I don't mention it, it's just as fine for me.